### PR TITLE
Document that our code is mostly not async cancellation-safe.

### PR DESCRIPTION
**Pull request by hlinnaka**
We had a hot debate on whether we should try to make our code cancellation-safe, or just accept that it's not, and make sure that our Futures are driven to completion. The decision is that we drive Futures to completion. This documents the decision, and summarizes the reasoning for that.

Discussion that sparked this:
https://github.com/neondatabase/neon/pull/4198#discussion_r1190209316

### CODE REVIEW

File: docs/pageserver-thread-mgmt.md [hunk 1]
```diff
@@ -21,19 +26,82 @@ also a `shudown_watcher()` Future that can be used with `tokio::select!`
...
+one of the Futures returns, and drops the others. In the Rust
+ecosystem, some functions are cancellation-safe, meaning they can be
+safely dropped without side-effects, while others are not. See
+documentation of `tokio::select!` for examples.
+
+In the pageserver and safekeeper, async code is *not*
+cancellation-safe by default. You can assume that any async code you
+write is polled to completion. There are a few exceptions to this,
+like the code that handles incoming HTTP requests, because Hyper will
+drop the request-handling Future if the HTTP connection is lost. The
```
Comment (hlinnaka) [inline docs/pageserver-thread-mgmt.md:42]: We haven''t actually done that "spawn a new task for each request" yet, but we should

File: docs/pageserver-thread-mgmt.md [hunk 2]
```diff
@@ -21,19 +26,80 @@ also a `shudown_watcher()` Future that can be used with `tokio::select!`
...
-The core storage code in `layered_repository/` is synchronous, with
-blocking locks and I/O calls. The current model is that we consider
-disk I/Os to be short enough that we perform them while running in a
-Tokio task. If that becomes a problem, we should use `spawn_blocking`
-before entering the synchronous parts of the code, or switch to using
-tokio I/O functions.
+It reads messages from one channel, sends them to another channel. If
+this code is cancelled at the `tx.send(i).await`, the message read
+from the receiver is lost. That may or may not be OK, depending on the
+context.
```
Comment (problame) [inline docs/pageserver-thread-mgmt.md:100]: rx and tx close together makes it seem like these are for the same channel.

Let's call `rx` the `work_inbox` and `tx` the `results_outbox`.
To make it clear these are separate channels.

File: docs/pageserver-thread-mgmt.md (hunk 1 above)
Comment (problame) [reply docs/pageserver-thread-mgmt.md:42]: Let's fix that before we merge this PR, or, create an issue and link to it from here.

**Review by problame — APPROVED** [review]
Fair description of the status quo.

**Review by koivunej — CHANGES_REQUESTED** [review]
We were discussing about this today.

I think it's a good idea to document that most of the code is not cancellation safe. 

I still think that it's a healthier option to use an approach like #4194 to build explicitly this with mutual exclusion, retrying, instead of having all request handlers spawned which might cause some unexpected issues as well.

**Comment by hlinnaka** [comment]
> We were discussing about this today.
> 
> I think it's a good idea to document that most of the code is not cancellation safe.
> 
> I still think that it's a healthier option to use an approach like #4194 to build explicitly this with mutual exclusion, retrying support, instead of having all request handlers spawned

Yeah, I know we won't agree unanimously on this.

> which might cause some unexpected issues as well.

Got some issues in mind that we should watch out for?

**Review by arssher — APPROVED** [review]
> Fair description of the status quo.

+1

I also don't see issues with plain spawns, not sure why an abstraction joining requests is wanted.

**Comment by hlinnaka** [comment]
Had to re-request review because github won't let push with "changes requested".

**Comment by koivunej** [comment]
From @hlinnaka:

> > I still think that it's a healthier option to use an approach like https://github.com/neondatabase/neon/pull/4194 to build explicitly this with mutual exclusion, retrying support, instead of having all request handlers spawned which might cause some unexpected issues as well.
>
> Got some issues in mind that we should watch out for?

Also @arssher:

> I also don't see issues with plain spawns, not sure why an abstraction joining requests is wanted.

Should we just add task spawning for each of the http handlers, it must change how a single threaded (go: blocking) client will observe things:

Before unconditional spawning:
```
  C: ------------X-------+-> time
 R1:         \----X  |   | 
 R2:                 \---/
```

In this art, R1 is the first timeouted or whatever cancelled, then R2 is it's second request.

Where X is timeout on the client side or any other reason to drop a connection, which could be a networking issue.

With spawning all request handlers:

```
  C: ------------X-------+---> time
 R1:         \----X  | ? |
 S1:          \------|-/ |
 R2:                 \---/
 S2:                  \-/
```

Now in this art, we have the first request (R1) similarly cancelled as before, handled to completion by first spawned task (S1). Second request is now started while the first one is still ongoing. I think the R2 response will be different than R1's would had been (the `?` on R1 line), and it's easy to see a situation where R2 will give a response of completion earlier than S1 completes.

The same situation with #4194:


```
  C: ------------X-------+---> time
 R1:         \----X  |   |
  S:          \------|-| |
 R2:                 \-|-/
```

Now here we only have one spawned task, which will be awaited by the R2 and then have it's result returned as the response. The same would happen with any number of concurrent requests.

I claim this is the safest alternative. Compared to this just spawning the request handlers might create more problems than what we hope to solve with it as there are other cases like:

```
  C: ------------X-------+---> time
 R1:         \----X  |   |  ?
 S1:          \------|---|--/
 R2:                 \---/
 S2:                  \-/
```

Should R2 report completion for example to the client would be very much wrong.

To summarize:
- I am not against documenting that most existing code might not be cancellation safe
- I am not against spawning tasks, that is the the way to avoid surprise cancellations
- I am against spawning all request handlers as tasks, and hoping it will decrease the problems we have

Now, this is what I was thinking over the weekend as a response. I haven't reviewed if this the spawning all request handlers is still in this PR. I'll approve if it is gone.

File: docs/pageserver-thread-mgmt.md [hunk 3]
```diff
@@ -21,19 +26,82 @@ also a `shudown_watcher()` Future that can be used with `tokio::select!`
...
-perform other long-running operations. For example, each WAL receiver
-connection is handled by a tokio Task. Once a piece of WAL has been
-received from the network, the task calls the blocking functions in
-the Repository to process the WAL.
+```
+while let Some(i) = work_inbox.recv().await {
+	if let Err(_) = results_outbox.send(i).await {
+		println!("receiver dropped");
+		return;
+		}
```
Comment (koivunej) [inline docs/pageserver-thread-mgmt.md:94]:
```suggestion
```

**Review by LizardWizzard — CHANGES_REQUESTED** [review]
> Hmm, maybe the issue here is that I didn't spell that out explicitly in the docs? 

I think so. I would also include the recommendation to annotate new functions as cancellation safe when they are. 

In general spawning is not enough, its normal for http or other networking frameworks to spawn tasks for every incoming request. If we unconditionally spawn our own task inside this results in 2x tasks causing the overhead that is not really needed. Tasks are cheap, but this is easily avoidable by SharedRetryable. So I dont really see the need in unconditional spawning.

I'll repeat my take on this from the comment:

If certain place relies on cancellation being safe -> use special primitive to be on the safe side and explicitly communicate the intention.

**Comment by LizardWizzard** [comment]
> Well, the alternative is that you cannot assume that. Which means that you need to make all code cancellation-safe.

Can you mark functions as cancel unsafe? And use SharedRetryable when needed? I think so. If you need to run cancel-unsafe things -> use shared retryable. This technically works without assuming that every piece of async code is polled to completion. And you dont have to make every piece of async code cancel safe.

I see this document as generally applicable for the whole code base and this claim is quite strong. I'd like to change the wording to not make such a strong claim. 

> If you use a function that can cause cancellation, you must be very careful to use only the functions that are specifically marked as cancellation-safe.

Rereading this, I realize that this directly violates the assumption. Assumtions that are immediately violated doesnt look useful to me and IMO are not in the spirit of the general project-wide documentation.

>  All our HTTP handlers can be cancelled, thanks to Hyper. Hence all HTTP handlers need protection from async cancellations.


I'd like to challenge that. Lets see what we have.

Lets see what we have.

GET "/v1/status", status_handler

No need to spawn.

PUT "/v1/failpoints", failpoints_handler

No need to spawn.

GET "/v1/tenant", tenant_list_handler

No need to spawn.

POST "v1/tenant", tenant_create_handler

Want to run this to completion, need request coalescing.

GET "/v1/tenant/:tenant_id" tenant_status

No need to spawn.

GET "/v1/tenant/:tenant_id/synthetic_size" tenant_size_handler

No need to spawn.

PUT "/v1/tenant/config" update_tenant_config_handler

Need request coalecsing. There shouldnt be concurrent modifications to config.

GET "/v1/tenant/:tenant_id/config" get_tenant_config_handler

No need to spawn.

GET "/v1/tenant/:tenant_id/timeline" timeline_list_handler

No need to spawn.

POST "/v1/tenant/:tenant_id/timeline" timeline_create_handler

Want to run to completion. Needs request coalescing.

POST "/v1/tenant/:tenant_id/attach" tenant_attach_handler

Want to run to completion. Needs request coalescing. Already uses spawn.

POST "/v1/tenant/:tenant_id/detach" tenant_detach_handler

Want to run to completion. Needs request coalescing.

POST "/v1/tenant/:tenant_id/load" tenant_load_handler

Want to run to completion. Needs request coalescing. Already uses spawn.

POST "/v1/tenant/:tenant_id/ignore" tenant_ignore_handler

Want to run to completion. Needs request coalescing.

GET "/v1/tenant/:tenant_id/timeline/:timeline_id" timeline_detail_handler

No need to spawn.

GET "/v1/tenant/:tenant_id/timeline/:timeline_id/get_lsn_by_timestamp" get_lsn_by_timestamp_handler

No need to spawn.

PUT "/v1/tenant/:tenant_id/timeline/:timeline_id/do_gc" timeline_gc_handler

No need to spawn. Gc needs to be idempotent. Needs request coalescing. Not sure how SharedRetryable will fit there, we already use mutex inside. Probably entire gc should be wrapped in shared retryable, not only http thing.

PUT "/v1/tenant/:tenant_id/timeline/:timeline_id/compact" timeline_compact_handler

The same as gc above

PUT "/v1/tenant/:tenant_id/timeline/:timeline_id/checkpoint" timeline_checkpoint_handler

The same as gc. Also all three are available under testing feature only. 

POST "/v1/tenant/:tenant_id/timeline/:timeline_id/download_remote_layers" timeline_download_remote_layers_handler_post

Needs to be idempotent, needs request coalescing. Already uses spawn. 

GET "/v1/tenant/:tenant_id/timeline/:timeline_id/download_remote_layers" timeline_download_remote_layers_handler_get

No need to spawn.

DELETE "/v1/tenant/:tenant_id/timeline/:timeline_id" timeline_delete_handler

Need to run to completion, needs request coalescing.

GET "/v1/tenant/:tenant_id/timeline/:timeline_id/layer" layer_map_info_handler

No need to spawn.

GET "/v1/tenant/:tenant_id/timeline/:timeline_id/layer/:layer_file_name" layer_download_handler

Implements request coalescing internally. Spawns.

DELETE "/v1/tenant/:tenant_id/timeline/:timeline_id/layer/:layer_file_name" evict_timeline_layer_handler

Needs request coalescing (takes appropriate locks). To my mind is cancel safe.

PUT "/v1/disk_usage_eviction/run" disk_usage_eviction_run

Needs request coalescing. Spawns internally.

These are obvious and used in tests only:

PUT "/v1/tenant/:tenant_id/break" handle_tenant_break
GET "/v1/panic" always_panic_handler
POST "/v1/tracing/event" post_tracing_event_handler

Some of the ones that are marked as "Want to run to completion" look cancel safe to me.

What is important? I dont see the case when we only need cancel safety without request coalescing.

This is the answer to:

> I want to separate the concerns of request coalescing and cancellation-safety. SharedRetryable addresses both, but the cancellation-safety issue is a wider concern. It's currently an issue with all HTTP handlers, not just with operations like timeline_deletion where SharedRetryable would be appropriate.

I'm against silver bullets (or things that pretend to be them). People need to know what cancellation is and how to deal with that. SharedRetryable is a useful primitive that solves the issues we have. 

I dont see why we need to spawn another task on every request given that it is not needed in many cases. If we spawn all handlers then with SharedRetryable it would be up to 3 tasks for every such request. I dont see why we need to do it this way.

Did I miss something?

File: docs/pageserver-thread-mgmt.md [hunk 4]
```diff
@@ -21,19 +26,82 @@ also a `shudown_watcher()` Future that can be used with `tokio::select!`
...
+In async Rust, futures can be "cancelled" at any await point, by
+dropping the Future. For example, `tokio::select!` returns as soon as
+one of the Futures returns, and drops the others. In the Rust
+ecosystem, some functions are cancellation-safe, meaning they can be
+safely dropped without side-effects, while others are not. See
+documentation of `tokio::select!` for examples.
+
+In the pageserver and safekeeper, async code is *not*
+cancellation-safe by default. You can assume that any async code you
+write is polled to completion. There are a few exceptions to this,
```
[the remark this answers is not in the document]
Comment (LizardWizzard) [reply docs/pageserver-thread-mgmt.md:40]: See https://github.com/neondatabase/neon/pull/4216#issuecomment-1548257457

**Comment by hlinnaka** [comment]
> > Well, the alternative is that you cannot assume that. Which means that you need to make all code cancellation-safe.
> 
> Can you mark functions as cancel unsafe? And use SharedRetryable when needed? I think so. If you need to run cancel-unsafe things -> use shared retryable. This technically works without assuming that every piece of async code is polled to completion. And you dont have to make every piece of async code cancel safe.
> 
> I see this document as generally applicable for the whole code base and this claim is quite strong. I'd like to change the wording to not make such a strong claim.

Can you propose some weaker phrasing that still captures the idea? How about this:

In the pageserver and safekeeper, async code is *not* cancellation-safe by default. Unless otherwise marked, any async functions that you call must be polled to completion, and conversely, you can assume that async functions that you write will be polled to completion. There are a few exceptions to this, ...

> > All our HTTP handlers can be cancelled, thanks to Hyper. Hence all HTTP handlers need protection from async cancellations.
> 
> I'd like to challenge that. Lets see what we have.
> 
> Lets see what we have.
> 
> <analysis that most of our HTTP handlers are in fact cancellations-safe>

The analysis for many of these depend on the fact that a lot of the underlying code is cancellation safe. So we would in fact need to analyse and mark a lot of underlying code as cancellation-safe. Including . e.g. `Timeline::get_reconstruct_data`. It's not async at the moment, but we're going in that direction with https://github.com/neondatabase/neon/pull/4220.

Is `refresh_gc_info_internal` cancellation-safe? I think it is, but it would be a little weird if it was cancelled: some timelines of the tenant might have their GC info updated, while others would not. That situation can also arise if the function returns early because of an error, but I don't think we expect errors here, that would be a symptom of some other issue, and at least you have a message in the log in that case.

Why would `refresh_gc_info_internal` need to be cancellation-safe? Because of the call chain:
tenant_size_handler -> gather_size_inputs -> gather_inputs -> refresh_gc_info.

I don't want to do this analysis on a case-by-case basis. It's tedious and error-prone. Even if we did all the work now, and committed to the ongoing effort in the future, I'd *still* want to have a wholesale solution to prevent from surprises and mistakes in the analysis. I'd still want to spawn a separate task for each request, just in case we missed something. There is no downside to spawning a task for each request.

**Comment by LizardWizzard** [comment]
> Can you propose some weaker phrasing that still captures the idea? How about this:

I'm against wrong assumptions. 

My approximate version:

In the pageserver and safekeeper, async code is not cancellation-safe by default for historical reasons. Unless otherwise marked, any async function that you call can violate cancellation safety. Thus using functions with primitives that rely on cancellation being safe should be done with extra care. Primitives include `select!` macro and `tokio::time::timeout`. Frequently in our codebase when cancellation-safety is needed, you also need to guard against concurrent access to some resources. Consider the timeout + retry example (Should be an examplanation). SharedRetryable/mutex/spawn, etc etc. 

> The analysis for many of these depend on the fact that a lot of the underlying code is cancellation safe. 

I think you fear cancelation safety too much. It is not a bug, its a feature that allows to avoid spending resources on things that are no longer needed. It is not that hard. Inspecting handlers that do get requests is really fast. In many cases it is enough to know that Mutex::lock is cancellation safe.

So I dont see why this is true:

> So we would in fact need to analyse and mark a lot of underlying code as cancellation-safe.

Management API is not that big. And you seem to completely ignore my argument that all http handlers we have that need cancellation safety need it together with request coalescing. You also ignored the point about spawning unnecessary tasks for each request. Again, cancellation is a feature, and we should embrace it instead of trying to get rid of it. We were moving from threads to also have better cancellation story in general and to avoid reinventing CHECK_FOR_INTERRUPTS for pageserver in particular.

> That situation can also arise if the function returns early because of an error

So you basically say that you need to be cancel-safe from the point of any return statement. Including question marks for early returns. How that is different from async cancellation safety? 

> I think it is, but it would be a little weird if it was cancelled

Weird or not, you showed that this can happen even without async context. Some functions need to be idempotent, and if they're not this is a bug.

> but I don't think we expect errors here

Thats the same for async cancellation, you dont expect for client requests to timeout, right? You cannot leave unhappy path aside. There is no discount.

> Why would refresh_gc_info_internal need to be cancellation-safe?

As we discussed above, this function needs to be cancel safe from the point of return statements too. timeline.update_gc_info can return an error and there is a question mark to return an error without finishing the loop. So in fact by making this function cancel safe from the point of return statement you'll make it async cancellation-safe too. 

Also I think the whole gc related API can be improved a lot so it is easier to work with it (and test it).

> I don't want to do this analysis on a case-by-case basis.

And you still need to. Any select/timeout call requires such analysis. The more you practice it - the faster you can do it. And when more code is cancel safe you dont need to check that many things for new code. Additionally this doesnt take into account SharedRetryable. If you need for something to be polled till the end -> use the right primitive.

Again, any return statement needs such analysis then too.
If you say that not any return statement is relevant, then the same is true for async cancellation. Some things are already cancel safe so you dont need to inspect them. Some tasks need to be idempotent regardless of async cancellation, so you cannot avoid the analysis.

> There is no downside to spawning a task for each request.

I think there is no point in doing that. Its just a waste of resources.

Again, we wanted to have cancellations when we were moving from threads, and now the idea is to fight them. I dont get it.

I think there was enough discussion. At this point we need to decide because it takes too much time without clear progress. We can set up a meeting to discuss that, or just vote here or just kill it.

**Comment by arssher** [comment]
> In the pageserver and safekeeper, async code is not cancellation-safe by default for historical reasons. 

I wouldn't call this 'historical reasons'. Cancellation safety is a separate feature, you may have it and you may not. Implementing it requires some care and effort, e.g. out of hand I can't usefully comment how hard is to have it for pageserver APIs, possibly not, but I don't know. Not having it and spawning almost doesn't need effort, but comes with obvious downsides, which doesn't seem to be an issue so far. Basically this PR just describes existing status quo, as Christian said. But it doesn't prevent us from adding cancellation safety with careful analysis once we have problems with current situation.
